src/overlays/actors/ovl_Door_Spiral/z_door_spiral.c

```diff
@@ -31,30 +40,221 @@ const ActorInit Door_Spiral_InitVars = {
+    play->doorCtx.transitionActorList[transitionActorId].id = -play->doorCtx.transitionActorList[transitionActorId].id;
+}
+
+void func_809A2DB0(DoorSpiral* this, PlayState* play) {
+    if (Object_IsLoaded(&play->objectCtx, this->objIndex) != 0) {
+        this->actor.objBankIndex = this->objIndex;
+        func_809A2B70(this, play);
+    }
+}
+
+f32 func_809A2E08(PlayState* play, DoorSpiral* this, f32 arg2, f32 arg3, f32 arg4) {
+    Player* player = GET_PLAYER(play);
+    Vec3f point;
+    Vec3f offset;
+
+    point.x = player->actor.world.pos.x;
+    point.y = player->actor.world.pos.y + arg2;
+    point.z = player->actor.world.pos.z;
+
+    Actor_OffsetOfPointInActorCoords(&this->actor, &offset, &point);
+
+    if ((arg3 < fabsf(offset.x)) || (arg4 < fabsf(offset.y))) {
+        return FLT_MAX;
+    } else {
+        return offset.z;
+    }
+}
+
+s32 func_809A2EA0(DoorSpiral* this, PlayState* play) {
+    Player* player = GET_PLAYER(play);
+
+    if (Player_InCsMode(play) == 0) {
+        SpiralInfo* temp_v0 = &sSpiralInfoTable[this->unk148];
+        f32 temp_fv0 = func_809A2E08(play, this, 0.0f, temp_v0->unk_0B, temp_v0->unk_0C);
+        if (fabsf(temp_fv0) < 64.0f) {
+            s16 var_v0 = player->actor.shape.rot.y - this->actor.shape.rot.y;
+            if (temp_fv0 > 0.0f) {
+                var_v0 = 0x8000 - var_v0;
+            }
+            if (ABS_ALT(var_v0) < 0x3000) {
+                return (temp_fv0 >= 0.0f) ? 1.0f : -1.0f;
+            }
+        }
+    }
+    return 0;
+}
+
+void func_809A2FF8(DoorSpiral* this, PlayState* play) {
+    Player* player;
+    u32 index;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Door_Spiral/func_809A2B70.s")
+    if (this->shouldClimb) {
+        func_809A2B60(this, func_809A3098);
+    } else if (func_809A2EA0(this, play) != 0) {
+        player = GET_PLAYER(play);
+        player->doorType = PLAYER_DOORTYPE_STAIRCASE;
+        player->doorDirection = this->unk146;
+        player->doorActor = &this->actor;
+        index = DOORSPIRAL_GET_FC00(&this->actor);
+        player->doorNext = (play->doorCtx.transitionActorList[index].params >> 0xA);
+        func_80122F28(player);
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Door_Spiral/func_809A2BF8.s")
+void func_809A3098(DoorSpiral* this, PlayState* play) {
+    Player* player = GET_PLAYER(play);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Door_Spiral/DoorSpiral_Init.s")
+    if (!(player->stateFlags1 & PLAYER_STATE1_20000000)) {
+        func_809A2B60(this, func_809A2DB0);
+        this->shouldClimb = 0;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Door_Spiral/DoorSpiral_Destroy.s")
+void DoorSpiral_Update(Actor* thisx, PlayState* play) {
+    DoorSpiral* this = THIS;
+    Player* player = GET_PLAYER(play);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Door_Spiral/func_809A2DB0.s")
+    if (!(player->stateFlags1 & 0x100004C0) || (this->actionFunc == func_809A2DB0)) {
```

Comment:
Could you fill in these stateflags?

Reply:
i don't really know how get it

Reply:
Oh, no worries. So the variable `player->stateFlags1` uses the macros `PLAYER_STATE1_` in `z64player.h`.

Then you want to break up `0x100004C0` into its individual bits. By convention, we try to go smallest to largest. So:
`0x100004C0` -> `0x40 | 0x80 | 0x400 | 0x10000000`
Does that step make sense to you?

Once you have the bits, you can then fill in the flags:
`0x40 | 0x80 | 0x400 | 0x10000000` -> `PLAYER_STATE1_40 | PLAYER_STATE1_80 | PLAYER_STATE1_400 | PLAYER_STATE1_10000000`